org.lflang/src/org/lflang/generator/LfExpressionVisitor.java

```diff
@@ -0,0 +1,160 @@
+/*
+ * Copyright (c) 2023, TU Dresden.
+ *
+ * Redistribution and use in source and binary forms, with or without modification,
+ * are permitted provided that the following conditions are met:
+ *
+ * 1. Redistributions of source code must retain the above copyright notice,
+ *    this list of conditions and the following disclaimer.
+ *
+ * 2. Redistributions in binary form must reproduce the above copyright notice,
+ *    this list of conditions and the following disclaimer in the documentation
+ *    and/or other materials provided with the distribution.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND ANY
+ * EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF
+ * MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL
+ * THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+ * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+ * PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT,
+ * STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF
+ * THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+package org.lflang.generator;
+
+import org.lflang.lf.BracedListExpression;
+import org.lflang.lf.Code;
+import org.lflang.lf.CodeExpr;
+import org.lflang.lf.Expression;
+import org.lflang.lf.LfFactory;
+import org.lflang.lf.Literal;
+import org.lflang.lf.ParameterReference;
+import org.lflang.lf.Time;
+
+/**
+ * A visitor for expressions in LF.
+ *
+ * @author Clément Fournier &lt;clement.fournier@tu-dresden.de&gt;
+ */
+public interface LfExpressionVisitor<P, R> {
```

Comment:
Cool! Do I understand that this is currently only used for copying expressions? I think this will become handy also for formatting expression and generating target code from expressions once the expression syntax becomes more complex.

Reply:
Yes currently it's used to replace parameter references with a value in the instance tree. The visiting and rebuilding is needed because a BracedListExpression is potentially recursive and can contain other list expressions. This will extend easily to other expressions, most of which are also recursive.